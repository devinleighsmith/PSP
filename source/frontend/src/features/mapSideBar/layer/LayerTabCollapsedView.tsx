import { groupBy } from 'lodash';
import React, { useEffect, useMemo } from 'react';
import { Col, Row } from 'react-bootstrap';
import { MdArrowLeft, MdArrowRight } from 'react-icons/md';

import { Button } from '@/components/common/buttons';
import TooltipIcon from '@/components/common/TooltipIcon';
import { exists } from '@/utils/utils';

import { InventoryTabNames } from '../property/InventoryTabs';
import { LayerContent } from './LayerContent';
import { LayerData } from './LayerTabContainer';

export interface ILayerTabCollapsedViewProps {
  layersData?: LayerData[];
  activeGroupedPages?: Record<string, number>;
  setActiveGroupedPages?: React.Dispatch<React.SetStateAction<Record<string, number>>>;
}

export interface TabLayerCollapsedView {
  content: React.ReactNode;
  key: InventoryTabNames;
  name: string;
}
/**
 * This view should be used to display multiple layers on a single screen stacked vertically, suitable for layers with limited data.
 * It should be used with the LayerData group property to create separate rows.
 */
export const LayerTabCollapsedView: React.FC<
  React.PropsWithChildren<ILayerTabCollapsedViewProps>
> = ({ layersData = [], activeGroupedPages = {}, setActiveGroupedPages }) => {
  const groupedLayersData = useMemo(() => groupBy(layersData, ld => ld.group), [layersData]);

  /** track the current page for each layer separately, ie layersData = [{..., group: 'group1' }, {..., group: 'group2' }] -> activeGroupedPages = {group1: 0, group2: 0} */
  useEffect(() => {
    const groupedPages = Object.keys(groupedLayersData).reduce((acc, key) => {
      acc[key] = 0; // start each group on page 0.
      return acc;
    }, {});
    setActiveGroupedPages(groupedPages);
  }, [groupedLayersData, setActiveGroupedPages]);

  return (
    <>
      {Object.keys(groupedLayersData).map(groupKey => (
        <React.Fragment key={groupKey}>
          <Row noGutters className="justify-content-around">
            <Col xs={1}>
              <Button
                variant="link"
                disabled={
                  !exists(activeGroupedPages[groupKey]) || activeGroupedPages[groupKey] === 0
                }
                onClick={() => {
                  setActiveGroupedPages({
                    ...activeGroupedPages,
                    [groupKey]: --activeGroupedPages[groupKey],
                  });
                }}
                title="previous page"
              >
                <MdArrowLeft size={36} />
              </Button>
            </Col>
            <Col xs="auto" className="d-flex align-items-center">
              <b>
                {groupedLayersData[groupKey][activeGroupedPages[groupKey]]?.title ??
                  groupedLayersData[groupKey][0]?.title}{' '}
                <TooltipIcon
                  toolTipId="layer-tab-tooltip"
                  toolTip="Lists all of the intersections of this layer and shape of the current parcel (see layers on map)."
                />
              </b>
            </Col>
            <Col xs={1} className="d-flex justify-content-end">
              <Button
                variant="link"
                disabled={
                  !exists(activeGroupedPages[groupKey]) ||
                  activeGroupedPages[groupKey] >= groupedLayersData[groupKey].length - 1
                }
                onClick={() => {
                  setActiveGroupedPages({
                    ...activeGroupedPages,
                    [groupKey]: ++activeGroupedPages[groupKey],
                  });
                }}
                title="next page"
              >
                <MdArrowRight size={36} />
              </Button>
            </Col>
          </Row>
          {exists(groupedLayersData[groupKey][activeGroupedPages[groupKey]]) ? (
            <LayerContent
              config={groupedLayersData[groupKey][activeGroupedPages[groupKey]]?.config}
              data={groupedLayersData[groupKey][activeGroupedPages[groupKey]]?.data}
            />
          ) : (
            <p>No Layer Data exists.</p>
          )}
        </React.Fragment>
      ))}
      {!exists(layersData?.length) || layersData.length === 0 ? <p>No Layer Data exists.</p> : null}
    </>
  );
};
