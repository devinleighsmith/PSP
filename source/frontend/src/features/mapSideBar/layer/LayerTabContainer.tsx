import { Feature, GeoJsonProperties } from 'geojson';
import { geoJSON, LatLngBounds } from 'leaflet';
import React, { useEffect, useMemo, useState } from 'react';

import { exists } from '@/utils';

import { ComposedProperty } from '../property/ComposedProperty';
import { InventoryTabNames } from '../property/InventoryTabs';
import {
  alrLayerConfig,
  electoralLayerConfig,
  firstNationsLayerConfig,
  getDynamicFeatureConfig,
  municipalityLayerConfig,
  parcelLayerConfig,
  pimsLayerConfig,
} from './constants';
import { ContentConfig } from './LayerContent';
import { ILayerTabCollapsedViewProps } from './LayerTabCollapsedView';
import { ILayerTabViewProps } from './LayerTabView';

export interface LayerData {
  title: React.ReactNode;
  bounds?: LatLngBounds;
  feature?: Feature;
  data: GeoJsonProperties;
  config: ContentConfig;
  tab: InventoryTabNames;
  group?: string;
}

type LayerTabViewComponent =
  | React.FunctionComponent<React.PropsWithChildren<ILayerTabViewProps>>
  | React.FunctionComponent<React.PropsWithChildren<ILayerTabCollapsedViewProps>>;

interface ILayerTabContainer {
  composedProperty: ComposedProperty | null;
  activeTab: InventoryTabNames;
  View: LayerTabViewComponent;
}

// Helper to build LayerData objects
function buildLayerData({
  features,
  titlePrefix,
  tab,
  config,
  dynamicConfig = false,
  totalCount,
  group,
}: {
  features?: Feature[];
  titlePrefix: string;
  tab: InventoryTabNames;
  config: ContentConfig;
  dynamicConfig?: boolean;
  totalCount?: number;
  group?: string;
}): LayerData[] {
  if (!exists(features) || !features?.length) return [];
  return features.map((feature, index) => ({
    title: (
      <>
        {titlePrefix} ({index + 1} of {features.length})
        {totalCount ? <i> out of {totalCount} total features</i> : ''}
      </>
    ),
    bounds: feature.geometry ? geoJSON(feature.geometry).getBounds() : undefined,
    data: feature.properties,
    feature,
    config: dynamicConfig ? getDynamicFeatureConfig(feature) : config,
    tab,
    group,
  }));
}

export const LayerTabContainer: React.FC<React.PropsWithChildren<ILayerTabContainer>> = ({
  composedProperty,
  activeTab,
  View,
}) => {
  const [activePage, setActivePage] = useState<number>(0);
  const [activeGroupedPages, setActiveGroupedPages] = useState<Record<string, number>>({});

  useEffect(() => {
    setActivePage(0); // reset the page whenever the tab changes.
    setActiveGroupedPages({}); // reset grouped pages as well
  }, [activeTab]);

  const layersData = useMemo(() => {
    if (!composedProperty) return [];
    const crownFeaturesTotal =
      (composedProperty.crownTenureFeatures?.length || 0) +
      (composedProperty.crownLeaseFeatures?.length || 0) +
      (composedProperty.crownInclusionFeatures?.length || 0) +
      (composedProperty.crownInventoryFeatures?.length || 0) +
      (composedProperty.crownLicenseFeatures?.length || 0);

    return [
      ...buildLayerData({
        features: composedProperty.pimsGeoserverFeatureCollection?.features,
        titlePrefix: 'PIMS data',
        tab: InventoryTabNames.pims,
        config: pimsLayerConfig,
      }),
      ...buildLayerData({
        features: composedProperty.parcelMapFeatureCollection?.features,
        titlePrefix: 'LTSA ParcelMap data',
        tab: InventoryTabNames.pmbc,
        config: parcelLayerConfig,
      }),
      ...buildLayerData({
        features: composedProperty.crownTenureFeatures,
        titlePrefix: 'Crown Land Tenures',
        tab: InventoryTabNames.crown,
        config: {},
        dynamicConfig: true,
        totalCount: crownFeaturesTotal,
      }),
      ...buildLayerData({
        features: composedProperty.crownLicenseFeatures,
        titlePrefix: 'Crown Land Licenses',
        tab: InventoryTabNames.crown,
        config: {},
        dynamicConfig: true,
        totalCount: crownFeaturesTotal,
      }),
      ...buildLayerData({
        features: composedProperty.crownLeaseFeatures,
        titlePrefix: 'Crown Land Tenures',
        tab: InventoryTabNames.crown,
        config: {},
        dynamicConfig: true,
        totalCount: crownFeaturesTotal,
      }),
      ...buildLayerData({
        features: composedProperty.crownInventoryFeatures,
        titlePrefix: 'Crown Land Inventory',
        tab: InventoryTabNames.crown,
        config: {},
        dynamicConfig: true,
        totalCount: crownFeaturesTotal,
      }),
      ...buildLayerData({
        features: composedProperty.crownInclusionFeatures,
        titlePrefix: 'Crown Land Inclusions',
        tab: InventoryTabNames.crown,
        config: {},
        dynamicConfig: true,
        totalCount: crownFeaturesTotal,
      }),
      ...buildLayerData({
        features: composedProperty.municipalityFeatures,
        titlePrefix: 'Municipality Information',
        tab: InventoryTabNames.other,
        config: municipalityLayerConfig,
        group: 'municipality',
      }),
      ...buildLayerData({
        features: composedProperty.alrFeatures,
        titlePrefix: 'ALR Information',
        tab: InventoryTabNames.other,
        config: alrLayerConfig,
        group: 'alr',
      }),
      ...buildLayerData({
        features: composedProperty.firstNationFeatures,

        titlePrefix: 'First Nations Information',
        tab: InventoryTabNames.other,
        config: firstNationsLayerConfig,
        group: 'first_nations',
      }),
      ...buildLayerData({
        features: composedProperty.electoralFeatures,
        titlePrefix: 'Electoral District Information',
        tab: InventoryTabNames.other,
        config: electoralLayerConfig,
        group: 'electoral',
      }),
      ...buildLayerData({
        features: composedProperty.highwayFeatures,
        titlePrefix: 'Highway Research',
        tab: InventoryTabNames.highway,
        config: {},
        dynamicConfig: true,
      }),
    ];
  }, [composedProperty]);

  const activeLayersData = layersData.filter(ld => ld.tab === activeTab);

  // Always pass all props, both views will ignore unused ones
  return (
    <View
      layersData={activeLayersData}
      activePage={activePage}
      setActivePage={setActivePage}
      activeGroupedPages={activeGroupedPages}
      setActiveGroupedPages={setActiveGroupedPages}
    />
  );
};
