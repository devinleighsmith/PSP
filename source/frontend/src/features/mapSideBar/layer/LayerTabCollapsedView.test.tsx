import React from 'react';
import { fireEvent, render } from '@/utils/test-utils';
import { LayerTabCollapsedView } from './LayerTabCollapsedView';
import { LayerData } from './LayerTabContainer';
import { vi } from 'vitest';

describe('LayerTabCollapsedView', () => {
  const setActiveGroupedPages = vi.fn();

  it('matches snapshot (empty layersData)', () => {
    const { container } = render(
      <LayerTabCollapsedView
        layersData={[]}
        activeGroupedPages={{}}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with no layers', () => {
    const { getByText } = render(
      <LayerTabCollapsedView
        layersData={[]}
        activeGroupedPages={{}}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(getByText(/no layer data exists/i)).toBeInTheDocument();
  });

  it('renders layers with no group', () => {
    const layersData: LayerData[] = [
      { title: 'Layer 1', data: {}, config: {}, tab: 'test' as any },
      { title: 'Layer 2', data: {}, config: {}, tab: 'test' as any },
    ];
    const { getByText, queryByText } = render(
      <LayerTabCollapsedView
        layersData={layersData}
        activeGroupedPages={{}}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(getByText('Layer 1')).toBeInTheDocument();
    expect(queryByText('Layer 2')).toBeNull();
  });

  it('renders layers with each layer having a unique group', () => {
    const layersData: LayerData[] = [
      { title: 'Layer 1', data: {}, config: {}, tab: 'test' as any, group: 'A' },
      { title: 'Layer 2', data: {}, config: {}, tab: 'test' as any, group: 'B' },
    ];
    const { getByText } = render(
      <LayerTabCollapsedView
        layersData={layersData}
        activeGroupedPages={{ A: 0, B: 0 }}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(getByText('Layer 1')).toBeInTheDocument();
    expect(getByText('Layer 2')).toBeInTheDocument();
  });

  it('renders layers with multiple layers in the same group', () => {
    const layersData: LayerData[] = [
      { title: 'Layer 1', data: {}, config: {}, tab: 'test' as any, group: 'A' },
      { title: 'Layer 2', data: {}, config: {}, tab: 'test' as any, group: 'A' },
      { title: 'Layer 3', data: {}, config: {}, tab: 'test' as any, group: 'B' },
    ];
    const { getByText, queryByText } = render(
      <LayerTabCollapsedView
        layersData={layersData}
        activeGroupedPages={{ A: 0, B: 0 }}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(getByText('Layer 1')).toBeInTheDocument();
    expect(queryByText('Layer 2')).toBeNull();
  });

  it('calls setActiveGroupedPages when right arrow clicked', () => {
    const layersData: LayerData[] = [
      { title: 'Layer 1', data: {}, config: {}, tab: 'test' as any, group: 'A' },
      { title: 'Layer 2', data: {}, config: {}, tab: 'test' as any, group: 'A' },
      { title: 'Layer 3', data: {}, config: {}, tab: 'test' as any, group: 'B' },
    ];
    const { getByText, getAllByTitle } = render(
      <LayerTabCollapsedView
        layersData={layersData}
        activeGroupedPages={{ A: 0, B: 0 }}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(getByText('Layer 1')).toBeInTheDocument();
    const rightButton = getAllByTitle('next page')[0];
    fireEvent.click(rightButton);
    expect(setActiveGroupedPages).toHaveBeenCalledWith({ A: 1, B: 0 });
  });

  it('disables right arrow when there is only one item in a group', () => {
    const layersData: LayerData[] = [
      { title: 'Layer 1', data: {}, config: {}, tab: 'test' as any, group: 'A' },
    ];
    const { getByText, getAllByTitle } = render(
      <LayerTabCollapsedView
        layersData={layersData}
        activeGroupedPages={{ A: 0, B: 0 }}
        setActiveGroupedPages={setActiveGroupedPages}
      />,
    );
    expect(getByText('Layer 1')).toBeInTheDocument();
    const rightButton = getAllByTitle('next page')[0];
    fireEvent.click(rightButton);
    expect(setActiveGroupedPages).not.toHaveBeenCalledWith({ A: 1 });
  });
});
