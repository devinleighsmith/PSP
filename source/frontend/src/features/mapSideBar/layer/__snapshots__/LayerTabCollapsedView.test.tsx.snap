// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LayerTabCollapsedView > matches snapshot (empty layersData) 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div />
  <p>
    No Layer Data exists.
  </p>
</div>
`;
